### PR TITLE
Add support for targetting language/country

### PR DESCRIPTION
**Pull request by shahthepro**
Adds methods to send push notification based country/language on codes and adds an endpoint to return all supported language and country codes.

### CODE REVIEW

File: infra/notifications/src/app.js [hunk 1]
```diff
@@ -471,6 +472,35 @@ app.post('/send_pn', pushAppAuth, async (req, res) => {
   res.send(data)
 })
 
+const cachedCountryCodes = []
+const cachcedLangCodes = []
+app.get('/push-targets', pushAppAuth, async (req, res) => {
+  if (!cachcedLangCodes.length) {
+    const [langCodes] = await db.sequelize.query(`
+      select distinct data->>'language' as langcode
```
Comment (franckc) [inline infra/notifications/src/app.js:480]: That's going to be a pretty slow query. An alternative would be to query the identity table on the 'country' column and add an index to it.

File: infra/notifications/src/app.js [hunk 2]
```diff
@@ -471,6 +472,35 @@ app.post('/send_pn', pushAppAuth, async (req, res) => {
   res.send(data)
 })
 
+const cachedCountryCodes = []
+const cachcedLangCodes = []
+app.get('/push-targets', pushAppAuth, async (req, res) => {
+  if (!cachcedLangCodes.length) {
```
Comment (franckc) [inline infra/notifications/src/app.js:478]: typo: cachedLangCodes

File: infra/notifications/src/app.js [hunk 3]
```diff
@@ -471,6 +472,35 @@ app.post('/send_pn', pushAppAuth, async (req, res) => {
...
+    const [langCodes] = await db.sequelize.query(`
+      select distinct data->>'language' as langcode 
+        from growth_participant order by data->>'language' ASC;
+    `)
+
+    cachcedLangCodes.push(...langCodes.map(l => l.langcode).filter(l => l))
+  }
+
+  if (!cachedCountryCodes.length) {
+    const [countryCodes] = await db.sequelize.query(`
```
Comment (franckc) [inline infra/notifications/src/app.js:488]: Pretty slow query. If performance becomes an issue, consider adding a DB index on the country column.

File: infra/notifications/src/mobilePush.js [hunk 4]
```diff
@@ -555,10 +455,86 @@ class MobilePush {
     await this._sendInChunks(data, deviceType, mobileRegisters)
   }
 
+  async _sendToCountryCode(data, deviceType, countryCode) {
```
Comment (franckc) [inline infra/notifications/src/mobilePush.js:458]: No need to fix now but maybe add a comment that if the server would get restarted in the middle of this method being called (for ex. due to a deploy), some notifications would not get sent.

File: infra/notifications/src/mobilePush.js [hunk 5]
```diff
@@ -555,10 +455,86 @@ class MobilePush {
...
+    `,
+      {
+        replacements: {
+          countryCode
+        }
+      }
+    )
+
+    if (!addresses.length) {
+      console.log('There are no addresses from', countryCode)
```
Comment (franckc) [inline infra/notifications/src/mobilePush.js:475]: nit: use logger

**Comment by shahthepro** [comment]
@franckc Thanks for the comments. yes regarding the queries, I did see the slowness on prod.  So, I just exported the values from there to static files: https://github.com/oplabs/pn-admin/tree/master/api/data

File: infra/notifications/src/mobilePush.js (hunk 4 above)
Comment (shahthepro) [reply infra/notifications/src/mobilePush.js:458]: Will add that.

File: infra/notifications/src/mobilePush.js [hunk 6]
```diff
@@ -555,10 +455,86 @@ class MobilePush {
...
+        deleted: false,
+        'permissions.alert': true
+      },
+      order: [['updatedAt', 'DESC']] // Most recently updated records first.
+    })
+
+    await this._sendInChunks(data, deviceType, mobileRegisters)
+  }
+
   async multicastMessage(data) {
```
Comment (franckc) [inline infra/notifications/src/mobilePush.js:532]: nit: in order to know what's going on on the server side, I would add a bit more logging throughout the code. at least something to track the progress of the notifications being sent. at least something like "Sending chunk X out of Y"

**Pull request merged**